### PR TITLE
Refactor scene initialization and continuations - Game_Temp 1 of 2

### PR DESCRIPTION
**Pull request by fmatthew5876**
This PR refactors the scene request / call mechanism. 

Fix: #1707

Objectives:
- [x] Fix edge cases in multiple scene calls
- [x] #1707 Test 22
- [x] #1707 Test 24
- [x] #1707 Test 25

### CODE REVIEW

**Comment by fmatthew5876** [comment]
This PR refactors the scene request API and then removes game temps for inn, shop, and name scene.

The last thing for `Game_Temp` is battle, but this is complicated and will require a lot of testing. These are pretty straightforward.

I'm putting the battle stuff in a follow up PR.

This has been tested and is ready for review and merge.

File: src/game_battle.h [hunk 1]
```diff
@@ -189,6 +192,13 @@ namespace Game_Battle {
 
 	/** @return death teleport handler if one is installed, otherwise an inactive target */
 	TeleportTarget GetDeathHandlerTeleport();
+
+	/** Don't reference this, use IsBattleRunning()! */
+	extern bool battle_running;
```
Comment (Ghabry) [inline src/game_battle.h]: Can't you move this in an anonymous namespace to prevent the access?

File: src/game_battle.h [hunk 2]
```diff
@@ -31,6 +31,9 @@ namespace RPG {
 }
 
 namespace Game_Battle {
+    /** @return true if a battle is currently running */
+    bool IsBattleRunning();
```
Comment (Ghabry) [inline src/game_battle.h]: wrong whitespace. Also considering that "Init" is the "constructor" it should be somewhere below Init

File: src/scene_shop.cpp [hunk 3]
```diff
@@ -16,25 +16,46 @@
...
+		std::vector<int> goods,
+		int shop_type,
+		bool allow_buy,
+		bool allow_sell,
+		Continuation on_finish)
+	: on_finish(std::move(on_finish))
+	  , goods(std::move(goods))
+	  , shop_type(shop_type)
+	  , allow_buy(allow_buy)
+	  , allow_sell(allow_sell)
```
Comment (Ghabry) [inline src/scene_shop.cpp]: the indent looks weird. 2 Tabs for the args and then "," in front for the init list

File: src/scene_shop.h [hunk 4]
```diff
@@ -37,12 +37,19 @@
...
 	/**
 	 * Constructor.
 	 */
-	Scene_Shop();
+	Scene_Shop(
+			std::vector<int> goods,
+			int shop_type,
+			bool allow_buy,
+			bool allow_sell,
+			Continuation on_finish);
```
Comment (Ghabry) [inline src/scene_shop.h]: same here

**Comment by Ghabry** [comment]
> Refactor requested scene API

Here it should be carefully checked that there are no side-effects by code that runs in the constructors. Before these happened in Push, now they happen when the scene is requested.

File: src/game_battle.h (hunk 1 above)
Comment (fmatthew5876) [reply src/game_battle.h]: Unfortunately not because it's in the header. I share your concerns, but I wanted to make sure access gets inlined.

This could be changed from namespace to struct and then it made private. But I did not want to make such a big change at this time.

File: src/scene_shop.cpp (hunk 3 above)
Comment (fmatthew5876) [reply src/scene_shop.cpp]: Fixed

File: src/scene_shop.h (hunk 4 above)
Comment (fmatthew5876) [reply src/scene_shop.h]: Fixed this one too

**Comment by fmatthew5876** [comment]
Addressed the review comments

> Here it should be carefully checked that there are no side-effects by code that runs in the constructors. Before these happened in Push, now they happen when the scene is requested.

Shop, Name, Save, Load, Gameover, and Menu are all affected in this PR and none of them do anything non-trivial in their constructors.

Battle does do actual game logic in it's constructor, but this is actually desired. 

As shown in #1642 when the battle scene is requested: RPG_RT plays battle start SFX, does battle transition, plays battle bgm, and clears the face. All of this happens even if you immediately cancel the battle scene by calling another scene in another parallel event on the same frame.

The clear face logic is implemented in #2027. The music/sfx logic is in #1202 

I have not implemented the special transition effect for cancelled scenes. This happens for both battle and gameover scene. Since it's only visual and such a wierd edge case (bug?) I'm not sure we care.

**Comment by Ghabry** [comment]
Jenkins: Test this please

File: src/game_battle.cpp [hunk 5]
```diff
@@ -68,7 +69,7 @@ void Game_Battle::Init() {
 	spriteset->Update();
 	animation.reset();
 
-	Game_Temp::battle_running = true;
+	Game_Battle::battle_running = true;
```
Comment (carstene1ns) [inline src/game_battle.cpp]: Why using the prefix here?

Comment (fmatthew5876) [reply src/game_battle.cpp]: More obvious it's a global variable
